src/realm/object-store/sectioned_results.cpp

```diff
@@ -364,16 +415,33 @@ void SectionedResults::calculate_sections()
 
 size_t SectionedResults::size()
 {
+    util::CheckedUniqueLock lock(m_mutex);
+    REALM_ASSERT(is_valid());
     calculate_sections_if_required();
     return m_sections.size();
 }
 
 ResultsSection SectionedResults::operator[](size_t idx)
 {
+    auto s = size();
+    REALM_ASSERT(idx < s);
```

Comment:
Bounds checking should throw an exception as it's not a bug in our code if the requested index is out of bounds.

Reply:
Changed the asserts to throw instead.